contracts/Orders.sol

```diff
@@ -0,0 +1,235 @@
+        }
+
+        // Update previous order to point to added order
+        prevOrder.nextId = orderId;
+
+        return orderId;
+    }
+
+    function getOrderDetails(uint32 orderId)
+        internal
+        view
+        returns (
+            address _maker,
+            uint128 _srcAmount,
+            uint128 _dstAmount,
+            uint32 _prevId,
+            uint32 _nextId
+        )
+    {
+        Order storage order = orders[orderId];
+        return (
+            order.maker,
+            order.srcAmount,
+            order.dstAmount,
+            order.prevId,
+            order.nextId
+        );
+    }
+
+    function removeById(uint32 orderId) internal {
+        verifyCanRemoveOrderById(orderId);
+
+        // Disconnect order from list
+        Order storage order = orders[orderId];
+        orders[order.prevId].nextId = order.nextId;
+        orders[order.nextId].prevId = order.prevId;
+
+        order.prevId = 0;
+        order.nextId = 0;
+    }
+
+    // The updated order id is returned following the update.
+    function update(uint32 orderId, uint128 srcAmount, uint128 dstAmount)
+        internal
+        returns(uint32)
+    {
+        address maker = orders[orderId].maker;
+        removeById(orderId);
+        return add(maker, srcAmount, dstAmount);
+    }
+
+    // The updated order id is returned following the update.
+    function updateWithPositionHint(
+        uint32 orderId,
+        uint128 srcAmount,
+        uint128 dstAmount,
+        uint32 prevId
+    )
+        internal
+        returns(uint32)
+    {
+        address maker = orders[orderId].maker;
+        removeById(orderId);
+        return addAfterId(maker, srcAmount, dstAmount, prevId);
+    }
+
+    function allocateIds(uint32 howMany) internal returns(uint32) {
+        uint32 firstId = nextId;
+        nextId += howMany;
+        return firstId;
+    }
+
+    function verifyCanRemoveOrderById(uint32 orderId) private view {
+        require(orderId != BUY_HEAD_ID);
+
+        Order storage order = orders[orderId];
+
+        // Make sure such order exists in mapping.
+        require(order.prevId != 0 || order.nextId != 0);
+    }
+
+    function calculateOrderSortKey(uint128 srcAmount, uint128 dstAmount)
+        public
+        pure
+        returns(uint)
+    {
+        return dstAmount * PRECISION / srcAmount;
+    }
+
+    function findPrevOrderId(
+        uint128 srcAmount,
+        uint128 dstAmount,
+        uint32 startId
+    )
+        public
+        view
+        returns(uint32)
+    {
+        uint newOrderKey = calculateOrderSortKey(srcAmount, dstAmount);
+
+        // This is okay for the HEAD ids, as their prevId is their id.
+        // TODO: rewrite in a simpler way - going to the prev of the provided
+        //       first id is not that elegant.
+        uint32 currId = orders[startId].prevId;
+        Order storage curr = orders[currId];
+        while (curr.nextId != TAIL_ID) {
+            currId = curr.nextId;
```

Comment:
I think u should first calculate key value for startID and compare. I assume this will serve adding with the "hinted" ID.
in this code, if first order key is smaller the new u can't find it.
meaning - if order should be before the start ID...